packages/button/src/Button.tsx

```diff
@@ -289,128 +288,94 @@ const disabledStyle = css`
-    ...(!usesLinkElement(props) && { disabled }),
-    'aria-disabled': disabled,
-  };
-
-  const rest = omit(props, [
-    'as',
-    'className',
-    'disabled',
-    'size',
-    'variant',
-    'children',
-    'glyph',
-  ]);
-
-  const spanStyle = css`
-    // Usually for this to take effect, you would need the element to be
-    /* ￿positioned￿. Due to an obscure part of CSS spec, flex children */
-    // respect z-index without the position property being set.
-    //
-    // https://www.w3.org/TR/css-flexbox-1/#painting
-    z-index: 1;
-    display: inline-flex;
-    flex-grow: 1;
-    justify-content: center;
-    align-items: center;
-  `;
-
-  const modifiedGlyph =
-    glyph && children
-      ? React.cloneElement(glyph, {
-          className: cx({ [glyphMargins[size]]: glyph != null }),
-        })
-      : glyph;
-
-  const renderButton = (Root: React.ElementType<any> = 'button') => (
-    <Root
-      ref={forwardRef}
-      type={Root === 'button' ? 'button' : undefined}
-      {...(rest as HTMLElementProps<any>)}
-      {...commonProps}
-    >
+      children,
+      disabled = false,
+      variant = Variant.Default,
+      size = Size.Normal,
+      glyph,
+      ...rest
+    }: BaseButtonProps,
+    ref: React.Ref<any>,
+  ) => {
+    const { usingKeyboard: showFocus } = useUsingKeyboardContext();
+
+    const commonProps = {
+      ref,
+      className: cx(
+        baseStyle,
+        buttonSizes[size],
+        buttonVariants[variant],
+        { [disabledStyle]: disabled },
+        { [focusStyle]: showFocus },
+        className,
+      ),
+      // only add a disabled prop if not an anchor
+      ...(rest.href === undefined && { disabled }),
+      'aria-disabled': disabled,
+    };
+
+    const spanStyle = css`
+      // Usually for this to take effect, you would need the element to be
+      /* ￿positioned￿. Due to an obscure part of CSS spec, flex children */
+      // respect z-index without the position property being set.
+      //
+      // https://www.w3.org/TR/css-flexbox-1/#painting
+      z-index: 1;
+      display: inline-flex;
+      flex-grow: 1;
+      justify-content: center;
+      align-items: center;
+    `;
+
+    const modifiedGlyph =
+      glyph && children
+        ? React.cloneElement(glyph, {
+            className: cx({ [glyphMargins[size]]: glyph != null }),
+          })
+        : glyph;
+
+    const content = (
       <span className={spanStyle}>
         {modifiedGlyph}
         {children}
       </span>
-    </Root>
-  );
-
-  if (usesCustomElement(props)) {
-    return renderButton(props.as);
-  }
+    );
 
-  if (usesLinkElement(props)) {
-    return renderButton('a');
-  }
+    if (rest.href !== undefined) {
+      return (
+        // we give button a default "as" value based on the `href` prop, if a custom
+        // "as" prop is supplied, it will overwrite this value through {...rest}
+        <Box as="a" {...commonProps} {...rest}>
```

Comment:
If there's a href won't Box handle setting it to an anchor element?

Reply:
the check is `as ? as : href ? : a : 'div'`

We need to supply an `as` when there is no href, but I suppose you're right and we can omit the explicit as setting on line 362.

Reply:
NVM! Ts evaluates the common props and sees attributes that don't exist on "div" tags, so it's expecting to receive a value for "as"

Reply:
This seems like a core part of the typing of Box that I'd expect to work. It shouldn't care about what a div can receive because href is set.

This might be an issue because based on the conditional – `rest.href !== undefined`. Based on that condition,`href` may still be falsy (or what TS cares about, not a string). If we check `typeof rest.href === 'string'` instead of whether `href` is undefined, that could resolve the issue.

Reply:
TS isn't understanding the check on line 358, I'm not sure why. If you inline href={rest.href} it works without defining an as..

Reply:
![Screen Shot 2020-06-23 at 5 07 30 PM](https://user-images.githubusercontent.com/26016393/85462931-0d2b5980-b574-11ea-9a32-c2b8cf34a058.png)

Reply:
👍